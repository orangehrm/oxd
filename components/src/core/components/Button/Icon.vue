<!--
/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->

<template>
  <button
    :disabled="disabled"
    v-if="withContainer"
    type="button"
    :class="classes"
    @click="onClick"
  >
    <oxd-icon :class="{'--disabled': disabled}" :name="name" />
  </button>
  <oxd-icon
    v-else
    :name="name"
    :class="{'oxd-icon-button__icon': true, '--disabled': disabled}"
    @click="onClick"
  />
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import Icon from '@ohrm/oxd/core/components/Icon/Icon.vue';

export default defineComponent({
  name: 'oxd-icon-button',

  components: {
    'oxd-icon': Icon,
  },

  props: {
    name: {
      type: String,
      required: true,
    },
    withContainer: {
      type: Boolean,
      default: true,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
  },

  emits: ['click'],

  computed: {
    classes(): object {
      return {
        'oxd-icon-button': true,
      };
    },
  },

  methods: {
    onClick(e: Event) {
      e.preventDefault();
      this.$emit('click', e);
    },
  },
});
</script>

<style src="./icon.scss" lang="scss" scoped></style>
