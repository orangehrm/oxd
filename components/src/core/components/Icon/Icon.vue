<!--
/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->

<script lang="ts">
import {computed, defineComponent, h, unref} from 'vue';
import {TYPES, TYPE_BOOTSTRAP} from './types';
import * as SVGs from './map';
import SVGLoader from '@ohrm/oxd/core/components/Icon/SVGLoader.vue';

export default defineComponent({
  name: 'oxd-icon',

  props: {
    name: {
      type: String,
      required: true,
    },
    type: {
      type: String,
      default: TYPE_BOOTSTRAP,
      validator: (value: string) => {
        return TYPES.indexOf(value) !== -1;
      },
    },
  },

  setup(props) {
    const classes = computed(() => {
      return {
        'oxd-icon': true,
        [`bi-${props.name}`]: props.type === TYPE_BOOTSTRAP,
      };
    });

    return () =>
      props.type === TYPE_BOOTSTRAP
        ? h('i', {class: unref(classes)})
        : h(
            SVGLoader,
            {class: unref(classes)},
            {default: () => h(SVGs[props.name])},
          );
  },
});
</script>

<style src="./icon.scss" lang="scss" scoped></style>
