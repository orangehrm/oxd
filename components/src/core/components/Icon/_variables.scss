$oxd-icon-font-color: $oxd-interface-gray-darken-1-color !default;
$oxd-icon-font-color--disabled: $oxd-interface-gray-lighten-2-color !default;

$oxd-icon-font-size--extra-large: 18px !default;
$oxd-icon-font-size--large: 16px !default;
$oxd-icon-font-size--medium: 14px !default;
$oxd-icon-font-size--small: 12px !default;
$oxd-icon-font-size--extra-small: 10px !default;

$oxd-svg-icon-font-size--extra-large: 22px !default;
$oxd-svg-icon-font-size--large: 20px !default;
$oxd-svg-icon-font-size--medium: 18px !default;
$oxd-svg-icon-font-size--small: 16px !default;
$oxd-svg-icon-font-size--extra-small: 14px !default;
$oxd-svg-icon-font-size--xx-small: 12px !default;
$oxd-svg-icon-font-size--xxx-small: 10px !default;

$oxd-help-popup-icon-color: $oxd-interface-gray-color !default;
$oxd-help-popup-icon-hover-color: $oxd-primary-one-color !default;
$oxd-help-popup-icon-z-index: $oxd-tooltip-font-z-index !default;

$oxd-help-popup-icon-text-font-family: $oxd-font-family !default;
$oxd-help-popup-icon-text-font-size: 0.75rem !default;
$oxd-help-popup-icon-text-font-style: normal !default;
$oxd-help-popup-icon-text-font-weight: 600 !default;
$oxd-help-popup-icon-text-color: $oxd-interface-gray-darken-1-color !default;
$oxd-help-popup-icon-text-background: $oxd-white-color !default;

$oxd-help-popup-icon-text-max-width: 200px !default;
$oxd-help-popup-icon-text-padding: 0.8rem !default;
$oxd-help-popup-icon-text-border-radius: 1rem !default;
$oxd-help-popup-icon-text-box-shadow: 0 0 35px 0 rgba(0, 0, 0, 0.18) !default;
