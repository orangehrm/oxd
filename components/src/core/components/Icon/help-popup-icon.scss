@import '../../../styles';

.oxd-svg-icon {
  color: $oxd-interface-gray-color;
  &:hover {
    color: $oxd-primary-one-color;
  }
}

[help-tooltip] {
  position: relative;
  cursor: pointer;
}

/* Make help icon show over the tooltip */
[help-tooltip]:hover > svg {
  z-index: $oxd-tooltip-font-z-index;
}

/* Applies to all tooltips */
[help-tooltip]::after {
  text-transform: none;
  font-size: 0.625rem;
  font-style: normal;
  font-weight: 400;
  line-height: 1;

  pointer-events: none;
  position: absolute;
  display: none;
  opacity: 0;

  content: attr(help-tooltip);

  font-family: $oxd-font-family;
  text-align: left;

  max-width: 200px;
  width: max-content;
  overflow: auto;
  white-space: pre-wrap;
  padding: 0.8rem;
  border-radius: 1rem;
  box-shadow: 0 0 35px 0 rgba(0, 0, 0, 0.18);
  background: #fff;
  color: $oxd-interface-gray-darken-1-color;
  z-index: $oxd-tooltip-font-z-index;
  top: 50%;
  left: 50%;
}

/* Make the tooltips respond to hover */
[help-tooltip]:hover::after,
[help-tooltip]:focus-visible::after {
  display: block;
  animation: tooltips-horz 300ms ease-out forwards;
}

/* don't show empty tooltips */
[help-tooltip='']::after {
  display: none !important;
}

/* KEYFRAMES */
@keyframes tooltips-vert {
  to {
    opacity: .9;
  }
}

@keyframes tooltips-horz {
  to {
    opacity: .9;
  }
}
