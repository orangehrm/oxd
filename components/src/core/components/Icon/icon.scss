@import '../../../styles';
@import 'variables';

@import '~bootstrap-icons/font/bootstrap-icons.css';

.oxd-icon {
  color: $oxd-icon-font-color;
}

[class^='bi-']::before,
[class*=' bi-']::before {
  line-height: inherit;
}
