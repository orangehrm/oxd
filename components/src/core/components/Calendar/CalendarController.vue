<template>
  <div class="oxd-calendar-header">
    <oxd-icon
      name="oxd-arrow-left"
      size="extra-small"
      @click="gotoPreviousMonth"
    ></oxd-icon>
    <ul class="oxd-calendar-selector">
      <oxd-calendar-dropdown class="oxd-calendar-selector-month">
        <div class="oxd-calendar-selector-month-selected">
          <oxd-text tag="p">{{ months[modelValue.month] }}</oxd-text>
          <oxd-icon name="caret-down-fill" :with-container="false" />
        </div>
        <template v-slot:content>
          <li
            role="none"
            class="oxd-calendar-selector --month"
            :class="{'--selected': index === modelValue.month}"
            v-for="(month, index) in months"
            :key="`oxd-month-${index}`"
            tabindex="0"
            @click="onSelectMonth(index)"
            @keyup.enter="onSelectMonth(index)"
          >
            {{ month }}
          </li>
        </template>
      </oxd-calendar-dropdown>
      <oxd-calendar-dropdown class="oxd-calendar-selector-year">
        <div class="oxd-calendar-selector-year-selected">
          <oxd-text tag="p">{{ modelValue.year }}</oxd-text>
          <oxd-icon name="caret-down-fill" :with-container="false" />
        </div>
        <template v-slot:content>
          <li
            role="none"
            class="oxd-calendar-selector --year"
            :class="{'--selected': year === modelValue.year}"
            v-for="(year, index) in years"
            :key="`oxd-year-${index}`"
            tabindex="0"
            @click="onSelectYear(year)"
            @keyup.enter="onSelectYear(year)"
          >
            {{ year }}
          </li>
        </template>
      </oxd-calendar-dropdown>
    </ul>
    <oxd-icon
      name="oxd-arrow-right"
      size="extra-small"
      @click="gotoNextMonth"
    ></oxd-icon>
  </div>
</template>
<script lang="ts">
import {defineComponent} from 'vue';
import Text from '@orangehrm/oxd/core/components/Text/Text.vue';
import Icon from '@orangehrm/oxd/core/components/Button/Icon.vue';
import CalendarDropdown from '@orangehrm/oxd/core/components/Calendar/CalendarDropdown.vue';

export default defineComponent({
  name: 'oxd-calendar-controller',
  props: {
    modelValue: {
      type: Object,
      required: true,
    },
    years: {
      type: Array,
      required: true,
    },
    months: {
      type: Array,
      required: true,
    },
  },
  emits: ['update:modelValue'],
  components: {
    'oxd-text': Text,
    'oxd-icon': Icon,
    'oxd-calendar-dropdown': CalendarDropdown,
  },
  methods: {
    calculateMonth(value: number) {
      const {month, year} = this.modelValue;
      if (month + value > 11 || month + value < 0) {
        return {
          month: month + value > 11 ? 0 : 11,
          year: year + value,
        };
      } else {
        return {
          month: month + value,
          year,
        };
      }
    },
    gotoPreviousMonth() {
      const {month, year} = this.calculateMonth(-1);
      this.$emit('update:modelValue', {
        month,
        year,
      });
    },
    gotoNextMonth() {
      const {month, year} = this.calculateMonth(1);
      this.$emit('update:modelValue', {
        month,
        year,
      });
    },
    onSelectYear(year: number) {
      this.$emit('update:modelValue', {
        month: this.modelValue.month,
        year,
      });
    },
    onSelectMonth(month: number) {
      this.$emit('update:modelValue', {
        month,
        year: this.modelValue.year,
      });
    },
  },
});
</script>
