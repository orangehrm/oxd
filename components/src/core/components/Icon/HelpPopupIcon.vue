<template>
  <oxd-icon
    :size="ICON_SIZE_SMALL"
    name="oxd-help-popup"
    :help-tooltip="$vt(helpText)"
  />
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import Icon from '@orangehrm/oxd/core/components/Icon/Icon.vue';
import translateMixin from '@orangehrm/oxd/mixins/translate';
import {ICON_SIZE_SMALL} from '@orangehrm/oxd/core/components/Icon/types';

export default defineComponent({
  name: 'oxd-help-popup-icon',

  components: {
    'oxd-icon': Icon,
  },

  mixins: [translateMixin],

  data() {
    return {
      ICON_SIZE_SMALL,
    };
  },

  props: {
    helpText: {
      type: String,
      default: null,
    },
  },
});
</script>
<style src="./help-popup-icon.scss" lang="scss" scoped></style>
