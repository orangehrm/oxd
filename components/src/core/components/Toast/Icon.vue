<!--
/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->

<template>
  <div class="oxd-toast-icon-container">
    <div :class="iconWrapClasses">
      <oxd-icon :name="iconName" class="oxd-toast-icon" />
    </div>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import {toastMixin} from './toast-mixin';
import {ToastType, ICON_MAP} from './types';
import Icon from '@ohrm/oxd/core/components/Icon/Icon.vue';

export default defineComponent({
  name: 'OxdToastIcon',

  components: {
    'oxd-icon': Icon,
  },

  mixins: [toastMixin],

  computed: {
    iconWrapClasses(): object {
      return {
        'oxd-toast-icon-wrap': true,
        [`oxd-toast-icon-wrap--${this.type}`]: true,
      };
    },
    iconName(): string {
      return ICON_MAP[this.type as ToastType];
    },
  },
});
</script>

<style src="./icon.scss" lang="scss" scoped></style>
