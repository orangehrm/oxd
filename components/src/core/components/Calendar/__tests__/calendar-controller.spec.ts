/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

import {mount, shallowMount} from '@vue/test-utils';
import CalendarController from '@ohrm/oxd/core/components/Calendar/CalendarController.vue';
import Icon from '@ohrm/oxd/core/components/Button/Icon.vue';

describe('CalendarController.vue', () => {
  // https://github.com/jsdom/jsdom/issues/1695
  window.HTMLElement.prototype.scrollIntoView = function () {
    //do nothing
  };

  const months = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
  ];
  const years = [2019, 2020, 2021, 2022];
  const modelValue = {
    month: 2,
    year: 2020,
  };
  it('renders OXD CalendarController', () => {
    const wrapper = mount(CalendarController, {
      props: {
        months,
        years,
        modelValue,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should open month dropdown on click', async () => {
    const wrapper = mount(CalendarController, {
      props: {
        months,
        years,
        modelValue,
      },
    });
    await wrapper.find('.oxd-calendar-selector-month').trigger('click');
    expect(
      wrapper.findAll(
        '.oxd-calendar-selector-month .oxd-calendar-dropdown--option',
      ).length,
    ).toEqual(12);
  });
  it('should open year dropdown on click', async () => {
    const wrapper = mount(CalendarController, {
      props: {
        months,
        years,
        modelValue,
      },
    });
    await wrapper.find('.oxd-calendar-selector-year').trigger('click');
    expect(
      wrapper.findAll(
        '.oxd-calendar-selector-year .oxd-calendar-dropdown--option',
      ).length,
    ).toEqual(4);
  });
  it('should select month on click', async () => {
    const wrapper = mount(CalendarController, {
      props: {
        months,
        years,
        modelValue,
      },
    });
    await wrapper.find('.oxd-calendar-selector-month').trigger('click');
    await wrapper
      .find('.oxd-calendar-selector-month .oxd-calendar-dropdown--option')
      .trigger('click');
    expect(wrapper.emitted('update:modelValue')).toEqual([
      [{month: 0, year: 2020}],
    ]);
  });
  it('should select year on click', async () => {
    const wrapper = mount(CalendarController, {
      props: {
        months,
        years,
        modelValue,
      },
    });
    await wrapper.find('.oxd-calendar-selector-year').trigger('click');
    await wrapper
      .find('.oxd-calendar-selector-year .oxd-calendar-dropdown--option')
      .trigger('click');
    expect(wrapper.emitted('update:modelValue')).toEqual([
      [{month: 2, year: 2019}],
    ]);
  });
  it('should select next month on click right arrow', async () => {
    const wrapper = shallowMount(CalendarController, {
      props: {
        months,
        years,
        modelValue,
      },
    });
    await wrapper.findAllComponents(Icon)[1].trigger('click');
    expect(wrapper.emitted('update:modelValue')).toEqual([
      [{month: 3, year: 2020}],
    ]);
  });
  it('should select previous month on click left arrow', async () => {
    const wrapper = shallowMount(CalendarController, {
      props: {
        months,
        years,
        modelValue,
      },
    });
    await wrapper.findAllComponents(Icon)[0].trigger('click');
    expect(wrapper.emitted('update:modelValue')).toEqual([
      [{month: 1, year: 2020}],
    ]);
  });
  it('should select next year on click right arrow if current month is december', async () => {
    const wrapper = shallowMount(CalendarController, {
      props: {
        months,
        years,
        modelValue: {month: 11, year: 2020},
      },
    });
    await wrapper.findAllComponents(Icon)[1].trigger('click');
    expect(wrapper.emitted('update:modelValue')).toEqual([
      [{month: 0, year: 2021}],
    ]);
  });
  it('should select previous year on click left arrow if current month is january', async () => {
    const wrapper = shallowMount(CalendarController, {
      props: {
        months,
        years,
        modelValue: {month: 0, year: 2020},
      },
    });
    await wrapper.findAllComponents(Icon)[0].trigger('click');
    expect(wrapper.emitted('update:modelValue')).toEqual([
      [{month: 11, year: 2019}],
    ]);
  });
});
