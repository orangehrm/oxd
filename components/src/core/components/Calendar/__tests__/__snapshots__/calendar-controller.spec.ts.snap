// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarController.vue renders OXD CalendarController 1`] = `
<div class="oxd-calendar-header"><button type="button" class="oxd-icon-button oxd-calendar-header-arrow" flow="top"><i class="oxd-svg-icon oxd-svg-icon--extra-small"><svg version="1.1" id="Capa_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px" transform="rotate(90)" viewBox="0 0 451.847 451.847" style="enable-background:new 0 0 451.847 451.847;" xml:space="preserve">
        <g>
          <defs></defs>
          <path fill="currentColor" d="M225.923,354.706c-8.098,0-16.195-3.092-22.369-9.263L9.27,151.157c-12.359-12.359-12.359-32.397,0-44.751c12.354-12.354,32.388-12.354,44.748,0l171.905,171.915l171.906-171.909c12.359-12.354,32.391-12.354,44.744,0c12.365,12.354,12.365,32.392,0,44.751L248.292,345.449C242.115,351.621,234.018,354.706,225.923,354.706z"></path>
        </g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
      </svg></i></button>
  <ul class="oxd-calendar-selector">
    <li tabindex="0" class="oxd-calendar-selector-month" for="month">
      <div class="oxd-calendar-selector-month-selected">
        <p class="oxd-text oxd-text--p">March</p><i class="oxd-icon oxd-icon--medium bi-caret-down-fill oxd-icon-button__icon" flow="top"></i>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </li>
    <li tabindex="0" class="oxd-calendar-selector-year" for="year">
      <div class="oxd-calendar-selector-year-selected">
        <p class="oxd-text oxd-text--p">2020</p><i class="oxd-icon oxd-icon--medium bi-caret-down-fill oxd-icon-button__icon" flow="top"></i>
      </div>
      <transition-stub>
        <!--v-if-->
      </transition-stub>
    </li>
  </ul><button type="button" class="oxd-icon-button oxd-calendar-header-arrow" flow="top"><i class="oxd-svg-icon oxd-svg-icon--extra-small"><svg version="1.1" id="Capa_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px" transform="rotate(-90)" viewBox="0 0 451.847 451.847" style="enable-background:new 0 0 451.847 451.847;" xml:space="preserve">
        <g>
          <defs></defs>
          <path fill="currentColor" d="M225.923,354.706c-8.098,0-16.195-3.092-22.369-9.263L9.27,151.157c-12.359-12.359-12.359-32.397,0-44.751c12.354-12.354,32.388-12.354,44.748,0l171.905,171.915l171.906-171.909c12.359-12.354,32.391-12.354,44.744,0c12.365,12.354,12.365,32.392,0,44.751L248.292,345.449C242.115,351.621,234.018,354.706,225.923,354.706z"></path>
        </g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
        <g></g>
      </svg></i></button>
</div>
`;
